File: src/com/timepath/hl2/io/demo/GameEventMessageType.java
package com.timepath.hl2.io.demo;

import com.timepath.io.BitBuffer;

/**
 *
 * @author TimePath
 */
public enum GameEventMessageType {

    /**
     * A zero terminated string
     */
    STRING(1) {

            @Override
            public Object parse(BitBuffer bb) {
                return bb.getString();
            }

        },
    /**
     * Float, 32 bit
     */
    FLOAT(2) {

            @Override
            public Object parse(BitBuffer bb) {
                return bb.getFloat();
            }

        },
    /**
     * Signed int, 32 bit
     */
    LONG(3) {

            @Override
            public Object parse(BitBuffer bb) {
                return bb.getInt();
            }

        },
    /**
     * Signed int, 16 bit
     */
    SHORT(4) {

            @Override
            public Object parse(BitBuffer bb) {
                return bb.getShort();
            }

        },
    /**
     * Unsigned int, 8 bit
     */
    BYTE(5) {

            @Override
            public Object parse(BitBuffer bb) {
                return bb.getByte();
            }

        },
    /**
     * Unsigned int, 1 bit
     */
    BOOL(6) {

            @Override
            public Object parse(BitBuffer bb) {
                return bb.getBoolean();
            }

        },
    /**
     * Any data, but not networked to clients
     */
    LOCAL(7);

    GameEventMessageType(int i) {

    }

    static GameEventMessageType get(int i) {
        GameEventMessageType[] vals = values();
        if(i < 1 || i > vals.length) {
            return null;
        }
        return vals[i - 1];
    }

    public Object parse(BitBuffer bb) {
        return null;
    }

}

File: src/main/kotlin/com/timepath/hl2/io/demo/GameEventMessageType.kt
package com.timepath.hl2.io.demo

import com.timepath.io.BitBuffer

public enum class GameEventMessageType(private val i: Int, val parse: (BitBuffer) -> Any?) {

    /** Marks the end of an event description */
    END(0, { throw UnsupportedOperationException() }),
    /** A zero terminated string */
    STRING(1, { it.getString() }),
    /** Float, 32 bit */
    FLOAT(2, { it.getFloat() }),
    /** Signed int, 32 bit */
    LONG(3, { it.getInt() }),
    /** Signed int, 16 bit */
    SHORT(4, { it.getShort() }),
    /** Unsigned int, 8 bit */
    BYTE(5, { it.getByte() }),
    /** Unsigned int, 1 bit */
    BOOL(6, { it.getBoolean() }),
    /** Any data, but not networked to clients */
    LOCAL(7, { null });

    companion object {
        fun get(i: Int) = values().firstOrNull { it.i == i }
    }
}
